package/MDAnalysis/coordinates/H5MD.py

```diff
@@ -0,0 +1,537 @@
+        """opens the trajectory file using h5py library"""
+        self._frame = -1
+        if isinstance(self.filename, h5py.File):
+            self._file = self.filename
+        else:
+            if self.parallel:
+                self._file = h5py.File(self.filename, 'r',
+                                       driver=self.driver,
+                                       comm=self.comm)
+            else:
+                self._file = h5py.File(self.filename, 'r')
+        # pulls first key out of 'particles'
+        # allows for arbitrary name of group1 in 'particles'
+        self._particle_group = self._file['particles'][
+                               list(self._file['particles'])[0]]
+
+    def close(self):
+        """close reader"""
+        self._file.close()
+        ## need to change this so it doesnt close the file, since
+        ## file might not be reopened in the same away
+
+    @property
+    def has_positions(self):
+        """True if 'position' group is in trajectory."""
+        return self._has['position']
+
+    @has_positions.setter
+    def has_positions(self, value: bool):
+        self._has['position'] = value
+
+    @property
+    def has_velocities(self):
+        """True if 'velocity' group is in trajectory."""
+        return self._has['velocity']
+
+    @has_velocities.setter
+    def has_velocities(self, value: bool):
+        self._has['velocity'] = value
+
+    @property
+    def has_forces(self):
+        """True if 'force' group is in trajectory."""
+        return self._has['force']
+
+    @has_forces.setter
+    def has_forces(self, value: bool):
+        self._has['force'] = value
+
+    @property
+    def n_frames(self):
+        """number of frames in trajectory"""
+        for name, value in self._has.items():
+            if value:
+                return self._particle_group[name]['value'].shape[0]
+
+    def _reopen(self):
+        """reopen trajectory"""
+        if self.parallel:
+            self._frame = -1
+            self._read_next_timestep()
+        else:
+            self.close()
+            self.open_trajectory()
+
+    def _read_frame(self, frame):
+        """reads data from h5md file and copies to current timestep"""
+        try:
+            for name, value in self._has.items():
+                if value:
+                    myframe = self._particle_group[name]['step'][frame]
+                    break
+            else:
+                raise NoDataError("Provide at least a position, velocity"
+                                  " or force group in the h5md file.")
+        except ValueError:
+            raise IOError from None
+
+        self._frame = frame
+        ts = self.ts
+        particle_group = self._particle_group
+        ts.frame = frame
+
+        # this block populates the data dictionary
+        # DT not read
+        if 'observables' in self._file:
+            data = self._file['observables']
+            for name in self._data_keywords:
```

Comment:
What happens if the `name` is not in the observables? I doubt that all h5md files contain an observable lambda.

Furthermore, we should be able to copy _any_ observable into data. Perhaps you can analyze the h5md in init and then set up the data_keywords accordingly. Or just iterate over all “data.keys()” and copy each for each step.